src/servers/grpc_server.cc

```diff
@@ -292,6 +294,82 @@ GRPCServer::Create(
   return Status::Success;
 }
 
+Status
+GRPCServer::CreateUniqueEndpointPorts(
```

Comment:
You shouldn't need to create these new functions (or in main.cc). Just following the existing flow but pass in 4 ports instead of 1. Also, use std::vector for endpoint_names, endpoint_ports, and grpc_servers. We typically avoid raw arrays because they are not as safe.

grpc_servers should be std::vector<unique_ptr<GRPServer>>* since when returning values we typically pass in an address.